### PR TITLE
feat: allow `while(true)` in `no-constant-condition`

### PR DESCRIPTION
**Pull request by Tanujkanti4441**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**
`no-constant-condition`
**What change do you want to make (place an "X" next to just one item)?**

[ ] Generate more warnings
[x] Generate fewer warnings
[ ] Implement autofix
[ ] Implement suggestions

**How will the change be implemented (place an "X" next to just one item)?**

[ ] A new option
[x] A new default behavior
[ ] Other

**Please provide some example code that this change will affect:**

```js
/*eslint no-constant-condition: ["error"]*/

while (true) {
    doSomething();
};

```

**What does the rule currently do for this code?**
rule reports this condition by default

**What will the rule do after it's changed?**
rule will not report this case by default

#### What changes did you make? (Give an overview)
changed the value of the existing option `checkloops` to `all`, `allExceptWhileTrue` and `none`;
where  `allExceptWhileTrue` will be new default and disallow constant expressions in all loops except `While(true)`

#### Is there anything you'd like reviewers to focus on?
Fixes: #17807

### CODE REVIEW

File: lib/rules/no-constant-condition.js
Comment (aladdin-add) [inline lib/rules/no-constant-condition.js:34]: it's a breaking change. can we allow `true`(same as "all"), `false`(same as "none") as suggested in the issue?

**Review by aladdin-add — CHANGES_REQUESTED** [review]

File: lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js:34]: sorry, i don't get it properly! are you asking for to set the values of `all` and `none` to `true` and `false` respectively? if this is the case then if we see following line from the issue:
> So we have agreed to extend the checkLoops option with three new values: all (like true), none (like false), and allExceptWhileTrue which will behave like all but will not flag while (true). allExceptWhileTrue will be the new default.

then seems like these values are correct.

Comment (aladdin-add) [reply lib/rules/no-constant-condition.js:34]: In previous versions, users could config:
```js
rules: { "no-constant-condition": ["error", { "checkLoops": false }] }
```

but it's no longer allowed in the change - it's a breaking change.

Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js:34]: i got this one
>can we allow true(same as "all"), false(same as "none") as suggested in the issue?

but confused in this line.

Comment (fasttime) [reply lib/rules/no-constant-condition.js:34]: Indeed we can't remove support for boolean values in `checkLoops` because we must maintain compatibility with existing configurations that use `true` or `false`. We don't want this change to be a breaking one.

The idea was to add support for three new new values but still allow `true` and `false` as synonyms for `"all"` and `"none"` respectively. Apologies if the confusion arose from my wording.

Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js:34]: Ah! so if i summarize it then `checkloops` can be configured in two ways:

option already has this one
```js
rules: { "no-constant-condition": ["error", { "checkLoops": true }] }
or
rules: { "no-constant-condition": ["error", { "checkLoops": false }] }
```
and also can be configured in this way
```js
rules: { "no-constant-condition": ["error", { "checkLoops": "all" }] }
or
rules: { "no-constant-condition": ["error", { "checkLoops": "allExceptWhileTrue" }] }
or
rules: { "no-constant-condition": ["error", { "checkLoops": "none" }] }
```
it is like using `oneOf` shchema

please correct if i am wrong.

Comment (fasttime) [reply lib/rules/no-constant-condition.js:34]: > Ah! so if i summarize it then `checkloops` can be configured in two ways:
> 
> option already has this one
> 
> ```js
> rules: { "no-constant-condition": ["error", { "checkLoops": true }] }
> or
> rules: { "no-constant-condition": ["error", { "checkLoops": false }] }
> ```
> 
> and also can be configured in this way
> 
> ```js
> rules: { "no-constant-condition": ["error", { "checkLoops": "all" }] }
> or
> rules: { "no-constant-condition": ["error", { "checkLoops": "allExceptWhileTrue" }] }
> or
> rules: { "no-constant-condition": ["error", { "checkLoops": "none" }] }
> ```
> 
> it is like using `oneOf` shchema

That's correct👍🏻

Comment (aladdin-add) [reply lib/rules/no-constant-condition.js:34]: I think it's right! 👍 
Thanks to @fasttime for helping me explain, as a non-native English speaker, forgive my bad English! 😅

Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js]: Done! added both ways to configure `checkLoops`.

File: lib/rules/no-constant-condition.js
Comment (mdjermanovic) [inline lib/rules/no-constant-condition.js:41]:
```suggestion
                        enum: ["all", "allExceptWhileTrue", "none", true, false]
```

I think we can simplify the schema this way. We're already using this pattern in other rules, for example:

https://github.com/eslint/eslint/blob/113f51ec4e52d3082a74b9682239a6e28d1a70ee/lib/rules/no-mixed-spaces-and-tabs.js#L25-L29

File: lib/rules/no-constant-condition.js
Comment (mdjermanovic) [inline lib/rules/no-constant-condition.js:142]:
```suggestion
                if (node.test.type === "Literal" && node.test.value === true && checkLoops === "allExceptWhileTrue") {
```

It's preferrable to check the AST.

File: lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js]: Done!

File: lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js]: Done!

File: docs/src/rules/no-constant-condition.md
Comment (fasttime) [inline docs/src/rules/no-constant-condition.md:230]: Did you mean `"checkLoops": false`? Because `"checkLoops": "none"` is used in the previous block.

```suggestion
/*eslint no-constant-condition: ["error", { "checkLoops": false }]*/
```

File: docs/src/rules/no-constant-condition.md
Comment (fasttime) [inline docs/src/rules/no-constant-condition.md:203]:
```suggestion
Examples of **correct** code for when `checkLoops` is `"none"` or `false`:
```

File: docs/src/rules/no-constant-condition.md
Comment (fasttime) [inline docs/src/rules/no-constant-condition.md:129]: We typically use double quotes for string values.

```suggestion
* `"all"` - Disallow constant expressions in all loops.
* `"allExceptWhileTrue"` (default) - Disallow constant expressions in all loops except `while` loops with expression `true`.
* `"none"` - Allow constant expressions in loops.

Or instead you can set the `checkLoops` value to booleans where `true` is same as `"all"` and `false` is same as `"none"`.

Examples of **incorrect** code for when `checkLoops` is `"all"` or `true`:
```

File: docs/src/rules/no-constant-condition.md
Comment (Tanujkanti4441) [reply docs/src/rules/no-constant-condition.md]: Done!

File: tests/lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply tests/lib/rules/no-constant-condition.js:1]: Done! please check if this is what you suggested.

File: docs/src/rules/no-constant-condition.md
Comment (Tanujkanti4441) [reply docs/src/rules/no-constant-condition.md]: Done!

File: tests/lib/rules/no-constant-condition.js
Comment (aladdin-add) [inline tests/lib/rules/no-constant-condition.js:175]: as it's the new default, good to have a test like:
```js
{ code: "while(true);"}
```

File: lib/rules/no-constant-condition.js
Comment (aladdin-add) [inline lib/rules/no-constant-condition.js:48]: I like to normalize the bool-style to the new options, so the following code can be unchanged. something like:

```js
if(options.checkLoops === true) checkLoops = "all";
if(options.checkLoops === false) checkLoops = "none";
```

File: docs/src/rules/no-constant-condition.md
Comment (fasttime) [inline docs/src/rules/no-constant-condition.md:189]:
```suggestion
Example of **correct** code for when `checkLoops` is `"allExceptWhileTrue"`:
```

**Comment by fasttime** [comment]
Thanks! LGTM after the suggestions.

File: tests/lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply tests/lib/rules/no-constant-condition.js:175]: Added!

File: lib/rules/no-constant-condition.js
Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js]: Done!

File: docs/src/rules/no-constant-condition.md
Comment (Tanujkanti4441) [reply docs/src/rules/no-constant-condition.md]: Done!

File: docs/src/rules/no-constant-condition.md
Comment (Tanujkanti4441) [reply docs/src/rules/no-constant-condition.md]: Done!

**Review by aladdin-add — DISMISSED** [review]
LGTM, thanks! Would like another review before merging.

File: lib/rules/no-constant-condition.js
Comment (fasttime) [inline lib/rules/no-constant-condition.js:55]: Shall we use an `else if`? I think that makes the logic easier to follow, but it's just my feeling.

```suggestion
        } else if (options.checkLoops === false) {
```

Comment (Tanujkanti4441) [reply lib/rules/no-constant-condition.js]: Done!

**Review by fasttime — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
